Crashlytics/Crashlytics/Models/FIRCLSOnDemandModel.m

```diff
@@ -0,0 +1,193 @@
+             withDataCollectionEnabled:(BOOL)dataCollectionEnabled
+            usingExistingReportManager:(FIRCLSExistingReportManager *)existingReportManager {
+  // Record the exception model into a new report if there is unused on-demand quota. Otherwise,
+  // log the occurence but drop the event.
+  @synchronized(self) {
+    if (![self isQueueFull]) {
+      FIRCLSDataCollectionToken *dataCollectionToken = [FIRCLSDataCollectionToken validToken];
+      NSString *activeReportPath = FIRCLSExceptionRecordOnDemandModel(
+          exceptionModel, self.recordedOnDemandExceptionCount, self.droppedOnDemandExceptionCount);
+
+      // Only submit an exception report if data collection is enabled. Otherwise, the report
+      // is stored until send or delete unsent reports is called.
+      double uploadDelay = [self calculateUploadDelay:YES];
+      if (dataCollectionEnabled) {
+        if (activeReportPath) {
+          // dispatch after calculateUploadDelay then update bucket with time of upload
+          [existingReportManager handleOnDemandReportUpload:activeReportPath
+                                        dataCollectionToken:dataCollectionToken
+                                                  withDelay:uploadDelay
+                                                   asUrgent:YES];
+          FIRCLSInfoLog(@"Submitted an on-demand exception with delay %.20f", uploadDelay);
+          [self updateBucketsAndCountWithDelay:uploadDelay];
+          return YES;  // Recorded and submitted the exception.
+        } else {
+          FIRCLSInfoLog(@"Error recording on-demand exception");
+          return NO;  // Something went wrong when recording the exception, so we don't have a valid
+                      // path.
+        }
+      }
+      FIRCLSInfoLog(@"Submitted an on-demand exception with delay %d", uploadDelay);
+      [self updateBucketsAndCountWithDelay:uploadDelay];
+      return YES;  // Recorded exception but did not submit since data collection is disabled.
+    } else {
+      FIRCLSInfoLog(@"No available on-demand quota. Dropping report.");
+      self.droppedOnDemandExceptionCount++;
+      [self calculateUploadDelay:YES];
+      return NO;  // Didn't record or submit then exception because no quota was available.
+    }
+  }
+}
+
+// Helper methods
+
+- (double)calculateUploadDelay:(BOOL)shouldUpdate {
+  // TODO: should return 0 if queue is empty?
+
+  double calculatedStepDuration = [self calculateStepDuration:shouldUpdate];
+  double power = pow(self.baseExponent, calculatedStepDuration);
+  NSNumber *calculatedUploadDelay =
+      [NSNumber numberWithDouble:(SEC_PER_MINUTE / self.uploadRate) * power];
+  NSComparisonResult result =
+      [[NSNumber numberWithDouble:MAX_DELAY_SEC] compare:calculatedUploadDelay];
+  return (result == NSOrderedAscending) ? MAX_DELAY_SEC : [calculatedUploadDelay doubleValue];
+}
+
+- (double)calculateStepDuration:(BOOL)shouldUpdate {
```

Comment:
I cannot tell if you already handled this case, but be careful how you initialize self.currentStep and self.lastUpdated so that if nothing happens for a long time after the app initialized, then the first on-demand fatal comes through, that the queueNotFullDuration still gets calculated correctly.

Reply:
Nit: I think "step duration" is a bit ambiguous here. The method calculateStepDuration returns an exponent to use in a formula that gives a duration of time, and the member variable self.stepDuration is an actual duration of time. Maybe name this method just "calculateStep" ?